### ISSUE TITLE
fly execute hangs until registry error message pops up

### ISSUE THREAD

**Issue by deitch** [concourse/fly#140]
Running 2.5.1 against dockerized concourse, started via compose using the sample in the "Getting started" guide, but modified to compose v2.

When I run any execute, it is stuck for a long time on `initializing`, then:

```
initializing
resource script '/opt/resource/check []' failed: exit status 1

stderr:
failed to ping registry: 2 error(s) occurred:

* ping https: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 127.0.0.11:53: read udp 127.0.0.1:59809->127.0.0.11:53: read: connection refused
* ping http: Get http://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 127.0.0.11:53: read udp 127.0.0.1:45233->127.0.0.11:53: read: connection refused

errored
```

I `exec` into the `web` and `worker` containers. In both cases, I can resolve `registry-1.docker.io` and can `curl` `https://registry-1.docker.io/v2/` and `http://registry-1.docker.io/v2/`.

**Comment by vito** [comment]
Looks like your system has `127.0.0.11` in `/etc/resolv.conf` which won't be addressable from within Concourse's containers, as it refers to the host's local network, whereas each container has its own network namespace.

You can explicitly configure DNS servers using `--garden-dns-server` (passing the flag once per DNS server). e.g. `concourse worker ... --garden-dns-server 8.8.8.8 --garden-dns-server 8.8.4.4`

**Comment by deitch** [comment]
I had commented on this in another issue, but cannot find from my mobile. 

I did manage to figure it out and override using the (mostly undocumented and difficult to find) CONCOURSE_GARDEN_DNS_SERVER env var. 

So two questions:

1. Given how the runc containers are dependent directly on resolv.conf in the host and its non convertible networking structure, should this be documented?
2. How good is it to just copy resolv.conf blindly? Since we already set up a linux bridge with its own address space, should we not have an address on the bridge for DNS that maps to a dnsmasq or similar on the host? That appears to create a much more robust structure.

**Comment by vito** [comment]
Ah, I was gonna link you to a similar sounding discussion here (https://github.com/cloudfoundry/guardian/issues/42) but then that turned out to be you. 😛 

I have a tough time knowing where docs for these kinds of things should live since I'm not sure where people would look, and it largely depends on an external dependency, which I prefer to push improvements onto via saner defaults (perhaps like what you're proposing). That being said, I think as we see more examples like this we may be able to collect things into a 'networking docs' category somewhere, which would also include configuring proxies and such.

@deitch Would you be OK with closing this out and continuing discussion on the aforementioned Guardian issue? I feel that most progression will happen there (aside from docs if we end up wanting that), and this issue has at least served to provide information for Googlers.

**Comment by deitch** [comment]
Yeah that was the issue in Guardian that I couldn't find! Too funny. "hey Siri, can you help me find the issues I cannot?" "I am sorry, I cannot help you. Ask @vito!"

Here is my perspective, feel free to accept or ignore.

There are 3 classes of people here:

1. Users of guardian. Those see the lower-level infrastructure and the issues and understand them. They don't need anything extra here (but that issue still is an issue)
2. Experts in Concourse. They get how things work, so aren't too surprised by the issue, but know what to do when it shows up. These people actually are just a subset of the first class.,
3. Users of concourse. These people don't know or care about guardian, but would love to use Concourse. They then are blocked by it "just not working".

The third group are the people you want to fall in love with and adopt Concourse, but just don't get guardian. 

The fact that guardian has a limitation (a design flaw due to a mistaken assumption about DNS, but really just technical debt that can be resolved) means nothing to them. It just means, "I cannot use this, oh well". Those are the people for whom we need a clear and obvious, "limitations and how to get around them, and we are working on making it better."

That would be a real pity, because I think this is the cleanest CI I have seen in quite some time. 

End of my (hopefully useful) input. 

I just had a CI part of a project, I did not want to go down the hosted (circle/travis/shippable) route, tried (again) Jenkins and suffered the complexity consequences, loved LambCI but it doesn't work cleanly with docker, and just wanted to do Concourse. The two issues that stopped me were: `CONCOURSE_EXTERNAL_URL` and DNS.

> Would you be OK with closing this out and continuing discussion on the aforementioned Guardian issue? I

Sure. Actually, howsabout we close this but update the docs? I think this should go in 3 places:

1. [Installing](https://concourse.ci/installing.html) : where is should describe the issue and how to work around it
2. [Docker](https://concourse.ci/docker-repository.html) : because it is most common here (actually, with compose vc2, it *always* will occur)
3. [Architecture](https://concourse.ci/architecture.html) : because it is an architectural issue.

A more module approach might be to have a "Troubleshooting" page which describes this issue, why it happens, and how to fix it. Then each of the above can mention the issue and link to the part in "Troubleshooting".

If the docs were markdown I would set up a PR, but I just cannot learn YATE (Yet Another Tempting Engine). :-)

**Comment by chendrix** [comment]
Moved to concourse/concourse#1066
